var/spack/repos/builtin/packages/autoconf/package.py

```diff
@@ -35,7 +35,9 @@ class Autoconf(Package):
     version('2.69', '82d05e03b93e45f5a39b828dc9c6c29b')
     version('2.62', '6c1f3b3734999035d77da5024aab4fbd')
 
-    depends_on('m4', type='build')
+    # Note: m4 is not only a build-time dependency of autoconf. The m4
+    # executable is needed when autoconf runs.
+    depends_on('m4 @1.4.6:', type=('build', 'run'))
```

Comment:
I am sorry, i think i lead you to the wrong direction

> “run”: The dependency package is used by this package at run time. The dependency package will be added to both PATH and PYTHONPATH at run time, but not during build time. “link” and “run” are similar in that they both describe a dependency that exists when the package is used, but they differ in the mechanism: “link” is via shared libraries, and “run” via an explicit search.

looks like `build` type here will make `m4` available in path when `autoconf` is a `build` dependency in another package.

see http://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-specs

You can check `.spack/build.env` of any installed package which use `autoconf`.